### ISSUE TITLE
Client setup error: No matching device for app

### ISSUE THREAD

**Issue by gpeal** [facebook/flipper#536]
## 🐛 Bug Report
@passy I set up Flipper in our app. It worked on a (non-64 bit) emulator but when I run in on our kiosk-like hardware, Flipper shows a bar that says:
Client setup error: No matching device for app: PACKAGE_NAME
![image](https://user-images.githubusercontent.com/1307745/64188880-18b46500-ce28-11e9-84e6-4efd1b48a1c1.png)

Flipper can see it for its generic features like logs and CPU info. However, it doesn't connect to the FlipperClient.

## To Reproduce
I'm not sure. It runs fine on an emulator but not on our hardware.

## Expected Behavior
Flipper connects to the FlipperClient.

## Environment
Our Kiosk-like hardware runs Marshmallow.

Thanks for working on Flipper. It looks really awesome and we have some ideas for useful plugins that we can build already.

**Comment by jknoxville** [comment]
Thanks for the report @gpeal 

I've played a little with a system app to investigate this, and have got it working with a file permission change that I'll ship asap.

However I didn't run into this `run-as` problem you mention, so I expect it won't fix your issue. In my tests, I had no problem executing `run-as` as the system app.

I tried first installing a user app and then moving it into `/system/app` and rebooting. I also tried placing an apk directly into `/system/app` on a fresh AVD. Both tests using marshmallow.

Could it be due to a variation in the OS running on your hardware? I was using an emulator for this so it might be more permissive.  Or perhaps I'm not installing the app as a system app correctly?

As for the `run-as` problem, I expect we can run execute the shell as root using `su`, instead of the app, to get around it. Something like [this](https://pastebin.com/rYN34qCi) - but note that that isn't tested at all because I haven't reproduced the issue.

**Comment by gpeal** [comment]
@jknoxville Our app also has `android:sharedUserId="android.uid.system"`. That may be the difference here.

**Comment by jknoxville** [comment]
Ah, thats probably it. That gives me: INSTALL_FAILED_SHARED_USER_INCOMPATIBLE. 

Reading up, it sounds like I'd have to sign the app with the same key that signed the firmware, is that correct?

**Comment by gpeal** [comment]
@jknoxville Yeah, that's correct. I can't share our own hardware setup (obviously) but maybe this would help in testing it out on an emulator: https://stackoverflow.com/a/37586790/715633

Thank you for taking the time to look at this. It will be a huge help to us if it works! 😄

**Comment by jsbeckr** [comment]
We have the same issue with React Native 0.62.0 and Flipper 0.35.0.

Our company devices are managed by Kaspersky Endpoint Security which checks the app and does <something> with it (but apparently not copying it to `/system/app`). That results in two app versions: the normal one and the managed/secure one. 

Flipper can connect to the normal app but not to the managed one with the same error message as @gpeal has. 

Is there some workaround for this issue or maybe pointers where I could investigate more?

Thanks!

**Comment by alk831** [comment]
I have the same issue after upgrading React Native to `0.62.0` from `0.61.5` with Flipper `0.35.0`.

**Comment by sirius60111** [comment]
same here with Flipper `0.37.0`

**Comment by romreed** [comment]
same here with Flipper 0.40.0

**Comment by gpeal** [comment]
Please leave this open. It is still an issue.

**Comment by gpeal** [comment]
Please add the backlog label. This is still an issue.

**Comment by jknoxville** [comment]
We made a change in the latest version of Flipper (0.70.0) to fall back to using `su` when a device is rooted and run-as isn't working: https://github.com/facebook/flipper/commit/805071fa4913338343615b4326ef10d9c3b116b0

It wasn't done to solve this issue in particular, but there's a chance it fixes this at the same time. Having such involved repro steps, I haven't checked. If not, I think we'll be relying on community support to fix this. PR's are very much welcome!

**Comment by gpeal** [comment]
@jknoxville Unfortunately, it is still failing on 0.70.0
![image](https://user-images.githubusercontent.com/1307745/104518847-3d3c5780-55ad-11eb-83de-5fb719a1ffc7.png)

**Comment by gpeal** [comment]
This still requires attention

**Comment by gpeal** [comment]
This still requires attention.
